rust/noosphere-sphere/src/authority/escalate.rs

```diff
@@ -0,0 +1,85 @@
+use crate::{HasMutableSphereContext, SphereContext};
+use std::sync::Arc;
+
+use anyhow::{anyhow, Result};
+use async_trait::async_trait;
+use noosphere_core::{
+    authority::Author,
+    data::{Did, Link, MemoIpld, Mnemonic},
+};
+use noosphere_storage::Storage;
+use std::future::Future;
+use tokio::sync::Mutex;
+use ucan::crypto::KeyMaterial;
+
+#[allow(missing_docs)]
+#[cfg(not(target_arch = "wasm32"))]
+pub trait SphereAuthoritySend: Send {}
+
+#[cfg(not(target_arch = "wasm32"))]
+impl<T> SphereAuthoritySend for T where T: Send {}
+
+#[allow(missing_docs)]
+#[cfg(target_arch = "wasm32")]
+pub trait SphereAuthoritySend {}
+
+#[cfg(target_arch = "wasm32")]
+impl<T> SphereAuthoritySend for T {}
+
+/// Any wrapper over [SphereContext] that implements [SphereAuthorityEscalate]
+/// has the ability to do some work on the [SphereContext] using a higher-privilege
+/// credential (in most cases, the root sphere credential).
+#[cfg_attr(not(target_arch = "wasm32"), async_trait)]
+#[cfg_attr(target_arch = "wasm32", async_trait(?Send))]
+pub trait SphereAuthorityEscalate<C, K, S>
+where
+    C: HasMutableSphereContext<K, S>,
+    K: KeyMaterial + Clone + 'static,
+    S: Storage + 'static,
+{
+    /// Given the recovery mnemonic for a sphere and a callback, invoke the
+    /// callback with a mutable [SphereContext] wrapper. The optional return
+    /// value of the callback should be the latest version of the sphere after
+    /// the work of the callback is performed
+    async fn with_root_authority<F, Fut>(&mut self, mnemonic: &Mnemonic, callback: F) -> Result<()>
```

Comment:
What's the rational for a callback here as opposed to returning a root sphere context? Isolating scope/access?

Reply:
It's to encourage atomicity in the change being made with root authority. In the case of `SphereCursor`, it gives an opportunity to advance the cursor to a specific new version without the benefit of knowing what was done with the other `SphereContext`.